### DISCUSSION TITLE
best way to add sparse behavioral annotations of timeseries data

### DISCUSSION THREAD

**Discussion by vikramvijayan** [dandi/helpdesk#75] [Q&A]
Hi,

I'd like to add annotations of behaviors that occur at specific times in my data. I tried using a timeseries, but because there are often 0 or 1 annotations per file .nwb file, an error was thrown (when the timeseries only had one data point). Please let me know what the best option is for very sparse behavioral annotations. I'm using MatNWB to create these files.

**Comment by CodyCBakerPhD** [comment]
Hi there,

> an error was thrown (when the timeseries only had one data point)

What was the exact error you saw for this?


For behavioral events such as this, the NWB Extension https://github.com/rly/ndx-events is usually the best way to encode such information (though it is done in PyNWB, not MatNWB). If you wanted, you could write all your information except for the annotated behaviors in MatNWB and then [append the file](https://pynwb.readthedocs.io/en/dev/tutorials/general/file.html?highlight=append#appending-to-an-nwb-file) using the extension data types.


Otherwise, the core NWB schema would have you do what you are doing now with a TimeSeries: generally follow the structure of [this behavior tutorial](https://pynwb--1464.org.readthedocs.build/en/1464/tutorials/domain/plot_behavior.html#behavioralevents-storing-behavioral-events), but using the [MatNWB version](https://neurodatawithoutborders.github.io/matnwb/doc/+types/+core/BehavioralEvents.html) of the container.


Best,
Cody Baker, PhD
Sr. Neuro-data Scientist at CatalystNeuro

**Reply by vikramvijayan** [reply]
Thanks Cody. 

Using the nwbinspector I get TypeError: TimeSeries.__init__: incorrect type for 'data' (got 'float64', expected 'ndarray, list, tuple, Dataset, StrDataset, HDMFDataset, AbstractDataChunkIterator, DataIO or TimeSeries'), incorrect type for 'timestamps' (got 'float64', expected 'ndarray, list, tuple, Dataset, StrDataset, HDMFDataset, AbstractDataChunkIterator, DataIO or TimeSeries')"

This is because I made a timeseries with only 1 entry in those 13 files.

Is there a way I could force the correct type in MatNWB even though there is just one element?

When I try to upload these files to Dandi I get "Caught exception 13 out of 194 files were found not containing all necessary metadata".

Thanks,
Vikram

**Reply by CodyCBakerPhD** [reply]
> Using the nwbinspector I get TypeError: TimeSeries.init: incorrect type for 'data' (got 'float64', expected 'ndarray, list, tuple, Dataset, StrDataset, HDMFDataset, AbstractDataChunkIterator, DataIO or TimeSeries'), incorrect type for 'timestamps' (got 'float64', expected 'ndarray, list, tuple, Dataset, StrDataset, HDMFDataset, AbstractDataChunkIterator, DataIO or TimeSeries')"

Can you upload the entire report from the NWBInspector? (from CLI, `--report-file-path some_file_name.txt`). From the sounds of it this is a validation level error; interesting that MatNWB does not throw an error about those fields when they are specified as scalars.

Another thing to try there would be to set the data/timestamps fields to arrays of only a single element - note that those fields of TimeSeries must be array-like, not single values (even if you only have a single value, or possibly no value at all, it still must be encapsulated in an array of some sort).

**Reply by vikramvijayan** [reply]
Thanks for such prompt responses. I'm attaching the report:
[nwb_inspector_log.txt](https://github.com/dandi/helpdesk/files/8718618/nwb_inspector_log.txt)

In Matlab I don't think there is a way to set the Timeseries (both data and timestamp) to an array unless I just add fake elements (like NaN or 0's)?

**Reply by CodyCBakerPhD** [reply]
@vikramvijayan Thanks for the full report. Indeed, MatNWB is producing unreadable NWBFiles in this case.

I'm not an expert in MatNWB so I'm not sure what the exact working syntax might be. I would have guessed doing `data=[some_value]`, but I do see how that does not differ in MATLAB from `data=some_value`. In PyNWB this is certainly doable, for example,

```
from pynwb import TimeSeries

example_time_series = TimeSeries(name="test", unit="test", data=["test"], timestamps=[2.3])
```

 I'd suggest raising an issue on https://github.com/NeurodataWithoutBorders/matnwb/issues/new/choose to ask them what they would suggest, and point out a minimal example of how what you're doing cannot then be read back (of course they may determine the fault lies with the PyNWB reader, but at least this gets it into their ecosystem).

**Reply by vikramvijayan** [reply]
Thanks. I issued a bug report and will see what options there are. If anyone else has ideas I am all ears!

Vikram
